sqldelight-runtime/src/test/java/com/squareup/sqldelight/SqlDelightDatabaseTest.kt

```diff
@@ -1,76 +1,101 @@
+  private lateinit var database: SqlDelightDatabase
+  private lateinit var connection: SqlDatabaseConnection
 
-    override fun onUpgrade(db: SupportSQLiteDatabase?, oldVersion: Int, newVersion: Int) = Unit
+  @Before fun setup() {
+    val databaseHelper = SqliteJdbcOpenHelper()
+    database = object : SqlDelightDatabase(databaseHelper) {}
+    connection = databaseHelper.getConnection()
   }
 
-  private val configuration = SupportSQLiteOpenHelper.Configuration.builder(RuntimeEnvironment.application)
-      .callback(callback)
-      .build()
-
-  private val helper = FrameworkSQLiteOpenHelperFactory().create(configuration)
-  private val database = object : SqlDelightDatabase(helper) { }
+  @After fun teardown() {
+    database.close()
+  }
 
-  @Test fun afterTransactionRunsAfterSuccessfulTransactionEnds() {
+  @Test fun `afterTransaction runs after transaction commits`() {
     val counter = AtomicInteger(0)
-    database.newTransaction().use {
+    database.transaction {
       database.afterTransaction { counter.incrementAndGet() }
       assertThat(counter.get()).isEqualTo(0)
-      it.markSuccessful()
-      assertThat(counter.get()).isEqualTo(0)
     }
 
     assertThat(counter.get()).isEqualTo(1)
   }
 
-  @Test fun afterTransactionDoesNotRunAfterUnsuccessfulTransactionEnds() {
+  @Test fun `afterTransaction does not run after transaction rollbacks`() {
     val counter = AtomicInteger(0)
-    database.newTransaction().use {
+    database.transaction {
       database.afterTransaction { counter.incrementAndGet() }
       assertThat(counter.get()).isEqualTo(0)
+      rollback()
     }
 
     assertThat(counter.get()).isEqualTo(0)
   }
 
-  @Test fun afterTransactionRunsImmediatelyWithNoTransaction() {
+  @Test fun `afterTransaction runs immediately with no transaction`() {
     val counter = AtomicInteger(0)
     database.afterTransaction { counter.incrementAndGet() }
     assertThat(counter.get()).isEqualTo(1)
   }
 
-  @Test fun afterTransactionRunsAfterParentTransactionEnds() {
+  @Test fun `afterTransaction runs after parent transaction commits`() {
     val counter = AtomicInteger(0)
-    database.newTransaction().use {
+    database.transaction {
       database.afterTransaction { counter.incrementAndGet() }
       assertThat(counter.get()).isEqualTo(0)
 
-      database.newTransaction().use {
+      database.transaction {
         database.afterTransaction { counter.incrementAndGet() }
         assertThat(counter.get()).isEqualTo(0)
-        it.markSuccessful()
-        assertThat(counter.get()).isEqualTo(0)
       }
 
       assertThat(counter.get()).isEqualTo(0)
-      it.markSuccessful()
-      assertThat(counter.get()).isEqualTo(0)
     }
 
     assertThat(counter.get()).isEqualTo(2)
   }
+
+  @Test fun `afterTransaction does not run in nested transaction when parent rolls back`() {
+    val counter = AtomicInteger(0)
+    database.transaction {
+      database.afterTransaction { counter.incrementAndGet() }
```

Comment:
I really think this should put the function on the transaction, not on the database

Reply:
... even if the database is only ever operated on a single thread, it still seems in the wrong scope

Reply:
the problem is we dont always know if we're in a transaction if we only have a reference to the database. It happens naturally in the library since we need to notify updated queries only after the transaction completes but if we're not in a transaction we can notify immediately

Reply:
I've made the transaction also subtype `Transactor` (renamed `SqlDelightDatabase`) so that you can at least call `afterCommit` and `afterRollback` without going through the database.